### DISCUSSION TITLE
May 2021 Priorities: Self-Hosted RTC, Handbook Structure, and Website

### DISCUSSION THREAD

**Discussion by jefftangx** [athensresearch/athens#1099] [Goals]
Legend:
- P1: high priority
- P2: medium priority
- P3: low priority

# Backend

<details>
<summary>P1: self-hosted RTC | Status: in progress | Next steps: make @arkRedM's multi-writer approach stable and/or attempt a single-writer approach</summary>

- [ ] @arkRedM's approach: multi-writer, use rule invariants to enforce implicit rules and fix bad data. Needs to heavily test to ensure no bugs.
- [ ] @neotyk and @tangjeff0's approach: centralize events to a single backend writer. Decouple events and stateful tx-data construction

</details>

<details>

<summary>P3: Spec out Homebase/Datahike Collaboration | Status: backlog | Reason: need a stable backend before we can use a Homebase PaaS | Next steps: discuss collaboration via contract work | Waiting on: them</summary>

- cc @smothers @rubinovitz
- This would allow users to easily spin up "on-prem" Athens for their team/org
- Open Questions (maybe should create separate discussion here)
  - [ ] Server-Sent Events vs WebSockets
  - [ ] Datascript vs Datahike vs Datalevin vs Crux vs Fluree (choose provider for both frontend and backend)
  - [ ] Cost/contract structure
</details>

# Community

<details>
<summary>P1: v1 Website | Status: in progress | Next steps: Shanberg and/or tangjeff0 make device-responsive</summary>

- [x] MVP content
- [ ] polish design
- link: https://athens-research.webflow.io/
- Related #509 , #1067 by @seekanddefine and #1072 by @limezy

</details>

<details>
<summary>P1: Gitbook Handbook Documentation | Status: in progress | Next steps: @seekanddefine creating 1-3 levels of structure for Handbook</summary>

- See #1078
- [ ] a table of contents and foundational structure for long-term docs at [Handbook](https://athensresearch.gitbook.io/handbook/).
  - This is what most of our Mid-Month Review covered. See retrospective on mix-up
- [x] Centralize all documentation -> Migrate ClojureFam / School of Athens docs to here and archive ClojureFam
</details>

<details>
<summary>Figure out differentiated strategy for Athens | Status: Done for now</summary>

1. focus on research, not "All in-one workspaces" for notetaking, task mgmt, project mgmt, etc.
2. focus on collaboration, first self-hosted then cloud-hosted

Goal could be renamed to "Figure out beachhead market and ideal customer persona, and orient product, marketing, and sales towards them."

</details>

# Frontend

<details>
<summary>No P1 priorities</summary>

## Notes from Mid-Point Sync

- core editing experience - most buggy from frontend POV
- multiplayer/presence UI - more aligned with remote dbs. could use help here to understand code
- db-picker - more aligned with remote dbs
- native toolbar - low priority but just want to close
- db-perf - junius working on it

## P2: core editing experience via ContentEditable and DB Performance

- [ ] migrate to a ContentEditable library for editing, e.g. ProseMirror
- [ ] improve performance for transacting blocks and search by optimizing datascript queries, moving off of posh, and properly deref'ing reagent atoms

## P3: general polish

Lots of other areas and issues for general improvements (like accessibility), but don't need to be done in any order.: https://www.craft.do/s/MAGVSD9LgwXSnF
</details>

**Comment by juniusfree** [comment]
> properly deref'ing reagent atoms

@tangjeff0 I'm learning reagent. Can I try this? What needs to get done here?

**Reply by jefftangx** [reply]
@juniusfree there are some places where we query the datascript db using posh, but we don't actually need to deref (and therefore don't need posh, which is known to be slow). Sometimes we do. I guess your job could be to identify where we use posh when we should use native datascript. @neotyk brought this to my attention so may add something.

Some but not all these queries are in `db.cljs`.

https://github.com/athensresearch/athens/blob/a81e12e7c981817960b03191b1040b8480be3fbe/src/cljs/athens/db.cljs#L267-L270

That said, I suspect there may be bigger performance gains by using datascript better. See this comment by @den1k https://github.com/athensresearch/athens/issues/756#issuecomment-831514372

**Reply by juniusfree** [reply]
@tangjeff0 I'll try to refactor some parts that use posh and then get your feedback when I'm done.

**Reply by juniusfree** [reply]
Not sure if this comment from tonksy is still relevant since this is made in 2015
> I don’t think it’s possible to have complex queries and high performance at the same time. When you need speed, use direct index lookup. https://github.com/tonsky/datascript/issues/130#issuecomment-159861446

**Reply by juniusfree** [reply]
It looks like `sort-block-children` makes the pages slow. I'll try if I can do something to fix it.

**Reply by juniusfree** [reply]
I can't find a significant difference in speed between posh and datascript. posh is even faster sometimes. Looking if there are other things that may affect speed.

**Reply by juniusfree** [reply]
I observed a delay in displaying the page _after_ querying. It's slower than the query.

**Reply by juniusfree** [reply]
Query is slower in the browser vs. repl 🤦‍♂️

**Reply by juniusfree** [reply]
https://github.com/athensresearch/athens/blob/a81e12e7c981817960b03191b1040b8480be3fbe/src/cljs/athens/views/pages/node_page.cljs#L541

This makes `node page` load twice everytime a page is opened

**Reply by juniusfree** [reply]
(Tried profiling in chrome for the first time) Both query and render are slow.

![image](https://user-images.githubusercontent.com/8164667/118384040-ad2f1d00-b635-11eb-9622-827ade138c0c.png)

**Reply by juniusfree** [reply]
Indexing is faster 😅

![image](https://user-images.githubusercontent.com/8164667/118384792-925fa700-b63b-11eb-8fbe-e06471f5eeb9.png)
![image](https://user-images.githubusercontent.com/8164667/118384806-a6a3a400-b63b-11eb-8c6e-42baeaa3447d.png)

**Reply by juniusfree** [reply]
indexing is faster on initial load. but posh is faster on re-render (e.g. opening a bullet, changing title)

![image](https://user-images.githubusercontent.com/8164667/118384962-3e55c200-b63d-11eb-8444-86c1a16fea40.png)

EDIT: because posh checks first for changes before the updating

**Reply by juniusfree** [reply]
removed posh from the codebase and updated with datascript equivalent functions. problem is reactivity.

**Reply by juniusfree** [reply]
@tangjeff0 https://github.com/athensresearch/athens/pull/1167 let me know if this is what needs to be done. Will update more later.

**Reply by jefftangx** [reply]
This looks great @juniusfree. Nice job using entity API. For the parts that need reactivity, we may need to create our own reagent wrapper around datascript.

Also I see you are logging time. There was some prior work here at #784

**Reply by juniusfree** [reply]
Thanks @tangjeff0. I'll check how that works.

I also copied this from that ticket. `(reagent/atom (d/empty-db schema) :meta {:listeners (atom {})})`. But I need to test more if this make the components reactive.
